ImageReader : Auto colorspace transforms should happen unpremulted

danieldresser-ie opened the pull request:
This might take some conversation, but it's a one line change, so I figured the conversation might as well happen on a PR.

@medubelko was having trouble getting PNGs loading into Gaffer to composite properly - for example, load `gaffer/graphics/boxOutNode.png` and then Merge it over white.

This is because we are automatically applying a linear -> sRGB conversion, but the conversion is being applied to a premultiplied image, producing dark fringes.  This PR applies the color space to an unpremulted image, producing a correct result.

My only concern with this is whether there is an extra cost of unpremulting and repremulting images where the color transform is an identity.  I don't see any obvious early out in OpenColorIOTransform to avoid pulling the colorData plug in cases where the input and output color spaces are the same - maybe I'm just missing something?  Even if we still need to add the early out though, it might be the right time to try and get correct behaviour for this?

johnhaddon commented:
LGTM, but not to Azure. Could you take a look at those test failures to see what's going on? Actually, I suspect we should be making the same change for ImageWriter as well, and that might explain some of it.

> My only concern with this is whether there is an extra cost of unpremulting and repremulting images where the color transform is an identity. I don't see any obvious early out in OpenColorIOTransform to avoid pulling the colorData plug in cases where the input and output color spaces are the same - maybe I'm just missing something?

I _think_ this is what `OpenColorIOTransform::enabled()` is achieving. Could you confirm?

danieldresser-ie commented:
Lets see, this ended up being a little bit complicated, because I had to adjust some tests.

Two tests had to have their thresholds raised very slightly.  I think this is innocuous - it looks like the thresholds were tuned to the exact values being written to file previously, so it's not surprising that changing the values would require a tiny adjustment.

I also had to adjust several of the expected result images in the test folder.  Most of these changes are very well justified.  In the tif and iff tests, a value of < 0.5, 0, 0, 0.5 > shouldn't be affected by the color space conversion - it's pure red with 50% coverage, and color transform doesn't affect pure red.  This is better than running the unpremultiplied value through a color transform, and getting < 0.75, 0, 0, 0.5 >, color brighter than pure red, as was happening previously.  The png and tga tests are even more obviously an improvement, since the files are stored on disk unpremultiplied, and these overbright values were actually getting clamped to 255 previously, preventing these images from roundtripping correctly ( if you load rgb.100x100.png into the Gaffer viewer in master, it does not match rgb.100x100.exr - this is now corrected with the new code and new expected images )

So that's a bit confusing, but a solid improvement.  The one place it feels a bit sketchy is with jpg - since we're discarding the alpha, is it better to prepare the image for proper alpha blending in jpeg color space, or is it conceptually better to think of it as comping the image onto a black background before the export, including before the color conversion ( which is equivalent to what we had before without the unpremult )?  At the end of the day, maybe we don't care too much about the precise colors you get when writing jpgs with alpha channels?  The overall advice would just be, "Don't do that."

> I think this is what OpenColorIOTransform::enabled() is achieving. Could you confirm?

Oh, missed that.  Yeah, that looks like it should be doing the right thing.  I can take a quick poke at it tomorrow to ensure it's actually working.

johnhaddon commented:
> I can take a quick poke at it tomorrow to ensure it's actually working.

I want to get this merged today, so I took a quick look myself. All is well, we had some test coverage for this already, and I've beefed it up further in b3698e7.

johnhaddon commented:
> The one place it feels a bit sketchy is with jpg - since we're discarding the alpha, is it better to prepare the image for proper alpha blending in jpeg color space, or is it conceptually better to think of it as comping the image onto a black background before the export, including before the color conversion ( which is equivalent to what we had before without the unpremult )?

Hmm. Having played with it (with a default Ramp), the new behaviour does seem less than ideal for JPG. I think I'd argue that by writing to a format without an alpha channel, I'm saying "I won't be comping this in future", so the best the ImageWriter can do is to comp on a black background before writing. That meets my intuition that when I reload that JPG, it should look the same in the Viewer.

> At the end of the day, maybe we don't care too much about the precise colors you get when writing jpgs with alpha channels? The overall advice would just be, "Don't do that."

I don't think this is limited to just JPG. What if I have an RGBA image, and I elect to write RGB-only to a format that could have supported alpha if I asked? And what if that format has a non-linear default space? Like DPX? Doesn't that sound like a comp delivery, and wouldn't we want to get that right by default?

johnhaddon commented:
> Note : I used GitHub's facility to push b3698e7 direct to this PR (into your branch) with the intention of merging immediately, then convinced myself that the no-alpha issue was actually an important one. So if you continue working, you'd want to `git pull origin imageReaderUnpremultColor` first.

danieldresser-ie commented:
> I don't think this is limited to just JPG. What if I have an RGBA image, and I elect to write RGB-only to a format that could have supported alpha if I asked? And what if that format has a non-linear default space? Like DPX? Doesn't that sound like a comp delivery, and wouldn't we want to get that right by default?

I did think about this case, and it is a bit ambiguous.  I concluded there was also a strong argument that if changing the channel mask from RGBA to RGB resulted in different data being written to the RGB channels, that could also be very confusing.  I feel like there's an expectation that removing A from the channel mask should just skip writing the channel, and not do anything else.

I'm not sure exactly what you're picturing for the practical case with a comp delivery ... I have an image with a meaningful alpha, and now for my comp delivery I want to omit the alpha channel, baking in a black background?  That doesn't sound very useful for anything.

Actually - I just realized an argument that I find very convincing that the DPX-with-a-channel mask case is currently correct, and it maybe persuades me that the jpg case is fine as well:  one of the most common cases for omitting the alpha would be if you are writing a stack of layers to separate files, and only include the alpha with one of them - this is how IE currently writes AOVs out of Arnold - we have beauty, specular, diffuse, etc. written to separate files, and we omit the alpha from all but the beauty, with the assumption that in comp you will reunite with channels of the AOVs with the alpha from the beauty.  This is a pretty reasonable workflow, and it would break in a very confusing way if we starting modifying the color space operations on AOVs where the alpha wasn't exported.

The analogy in the jpg case would be if you exported the alpha channel as a separate grayscale jpg.  The current approach would yield correct results if you then combined them back together in another program ( which I suppose someone may have done at some point in a game engine ).

Fundamentally, I don't feel like it's worth wasting too much time speculating about what someone wants if they write images with alpha masks to jpgs.  The assumption that they're going to export the alpha channel separately and then recombine seems as reasonable as any other - if they actually want it on a black background, it's easy for them to Merge it with opaque black, or delete the alpha channel.

If you want me to add in a special case for jpgs, I can do that, but I'm now pretty convinced that we shouldn't hack the DPX-with-RGB-channel-mask case.

danieldresser-ie commented:
Also: remaining test failure appears to be because Arnold version in the dependencies hasn't been updated yet.

johnhaddon commented:
> I concluded there was also a strong argument that if changing the channel mask from RGBA to RGB resulted in different data being written to the RGB channels, that could also be very confusing. I feel like there's an expectation that removing A from the channel mask should just skip writing the channel, and not do anything else.

Yeah, worded like that it sounds very reasonable. But there's also an expectation of round-tripping of visual results which is pretty reasonable and is currently broken. Perhaps there's a reason nobody complained about the old behaviour?

> Actually - I just realized an argument that I find very convincing that the DPX-with-a-channel mask case is currently correct, and it maybe persuades me that the jpg case is fine as well: one of the most common cases for omitting the alpha would be if you are writing a stack of layers to separate files, and only include the alpha with one of them - this is how IE currently writes AOVs out of Arnold - we have beauty, specular, diffuse, etc. written to separate files, and we omit the alpha from all but the beauty, with the assumption that in comp you will reunite with channels of the AOVs with the alpha from the beauty.

Does this actually work though? If we unpremult->transform->premult on write, we can't do the reverse unpremult->inversetransform->premult on read, because we don't have the alpha channel available at that point. To be symmetric, we need to ignore the alpha on write, because it won't be there on read. Maybe I'm missing something, but _I think_ you've just made a case for the behaviour I was suggesting. It's a bit hard to test because neither JPG or DPX will accept A on it's own, but ignoring that, here's a little test network. The right hand branch emulates what I think we should be doing, and appears to work...

<details>

```
import Gaffer
import GafferDispatch
import GafferImage
import IECore
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 56, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Ramp"] = GafferImage.Ramp( "Ramp" )
parent.addChild( __children["Ramp"] )
__children["Ramp"]["ramp"].clearPoints()
__children["Ramp"]["ramp"].addChild( Gaffer.ValuePlug( "p0", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Ramp"]["ramp"]["p0"].addChild( Gaffer.FloatPlug( "x", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Ramp"]["ramp"]["p0"].addChild( Gaffer.Color4fPlug( "y", defaultValue = imath.Color4f( 0, 0, 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Ramp"]["ramp"].addChild( Gaffer.ValuePlug( "p1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Ramp"]["ramp"]["p1"].addChild( Gaffer.FloatPlug( "x", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Ramp"]["ramp"]["p1"].addChild( Gaffer.Color4fPlug( "y", defaultValue = imath.Color4f( 0, 0, 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Ramp"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ImageReader"] = GafferImage.ImageReader( "ImageReader" )
parent.addChild( __children["ImageReader"] )
__children["ImageReader"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["CopyChannels"] = GafferImage.CopyChannels( "CopyChannels" )
parent.addChild( __children["CopyChannels"] )
__children["CopyChannels"]["in"].addChild( GafferImage.ImagePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["CopyChannels"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ImageWriter1"] = GafferImage.ImageWriter( "ImageWriter1" )
parent.addChild( __children["ImageWriter1"] )
__children["ImageWriter1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["DeleteChannels"] = GafferImage.DeleteChannels( "DeleteChannels" )
parent.addChild( __children["DeleteChannels"] )
__children["DeleteChannels"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ImageWriter2"] = GafferImage.ImageWriter( "ImageWriter2" )
parent.addChild( __children["ImageWriter2"] )
__children["ImageWriter2"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ImageReader1"] = GafferImage.ImageReader( "ImageReader1" )
parent.addChild( __children["ImageReader1"] )
__children["ImageReader1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["CopyChannels1"] = GafferImage.CopyChannels( "CopyChannels1" )
parent.addChild( __children["CopyChannels1"] )
__children["CopyChannels1"]["in"].addChild( GafferImage.ImagePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["CopyChannels1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Ramp"]["startPosition"].setValue( imath.V2f( 0, 540 ) )
__children["Ramp"]["endPosition"].setValue( imath.V2f( 1920, 540 ) )
__children["Ramp"]["ramp"]["p1"]["x"].setValue( 1.0 )
__children["Ramp"]["ramp"]["p1"]["y"].setValue( imath.Color4f( 1, 1, 1, 1 ) )
__children["Ramp"]["__uiPosition"].setValue( imath.V2f( -15.6491013, 5.45000029 ) )
__children["ImageReader"]["fileName"].setValue( '/tmp/rgb.jpg' )
__children["ImageReader"]["refreshCount"].setValue( 14 )
__children["ImageReader"]["__uiPosition"].setValue( imath.V2f( -51.9260292, -16.7101097 ) )
__children["CopyChannels"]["in"]["in0"].setInput( __children["ImageReader"]["out"] )
__children["CopyChannels"]["in"]["in1"].setInput( __children["Ramp"]["out"] )
__children["CopyChannels"]["channels"].setValue( '[A]' )
__children["CopyChannels"]["__uiPosition"].setValue( imath.V2f( -48.9260216, -27.6741734 ) )
__children["ImageWriter1"]["in"].setInput( __children["Ramp"]["out"] )
__children["ImageWriter1"]["fileName"].setValue( '/tmp/rgb.jpg' )
__children["ImageWriter1"]["channels"].setValue( '[RGB]' )
Gaffer.Metadata.registerValue( __children["ImageWriter1"]["channels"], 'channelMaskPlugValueWidget:custom', True )
__children["ImageWriter1"]["__uiPosition"].setValue( imath.V2f( -50.7583122, -8.98421478 ) )
__children["DeleteChannels"]["in"].setInput( __children["Ramp"]["out"] )
__children["DeleteChannels"]["channels"].setValue( '[A]' )
__children["DeleteChannels"]["__uiPosition"].setValue( imath.V2f( 7.04423761, -9.15341759 ) )
__children["ImageWriter2"]["in"].setInput( __children["DeleteChannels"]["out"] )
__children["ImageWriter2"]["fileName"].setValue( '/tmp/rgb2.jpg' )
__children["ImageWriter2"]["channels"].setValue( '[RGB]' )
Gaffer.Metadata.registerValue( __children["ImageWriter2"]["channels"], 'channelMaskPlugValueWidget:custom', True )
__children["ImageWriter2"]["__uiPosition"].setValue( imath.V2f( 5.54423761, -17.3174801 ) )
__children["ImageReader1"]["fileName"].setValue( '/tmp/rgb2.jpg' )
__children["ImageReader1"]["refreshCount"].setValue( 15 )
__children["ImageReader1"]["__uiPosition"].setValue( imath.V2f( 5.16927814, -23.4949112 ) )
__children["CopyChannels1"]["in"]["in0"].setInput( __children["Ramp"]["out"] )
__children["CopyChannels1"]["in"]["in1"].setInput( __children["ImageReader1"]["out"] )
__children["CopyChannels1"]["channels"].setValue( '[A]' )
__children["CopyChannels1"]["__uiPosition"].setValue( imath.V2f( 5.16927814, -31.6589737 ) )


del __children
```

</details>

To achieve the same on the left branch, I think you'd need to disable colour space conversions on the ImageReader, copy in the alpha, unpremult, manually apply the inverse transform and then premult again. Hopefully I'm not going crazy...

johnhaddon commented:
And here's a tiny little network which isn't totally implausible and gives surprising results for JPG currently. If getting this naive little case to work nicely requires the same as your more production-motivated case, then that's good enough for me (assuming I'm not wrong about the production case).

danieldresser-ie commented:
> Perhaps there's a reason nobody complained about the old behaviour?

I very much suspect no one has much used Gaffer with alpha channels with anything other than EXR.  As soon as Michael tried compositing with a png source in Gaffer, he noticed that he couldn't get his edges right.

> Does this actually work though? If we unpremult->transform->premult on write, we can't do the reverse unpremult->inversetransform->premult on read, because we don't have the alpha channel available at that point.

Hmm, that's a very fair point.  I was considering the case where you were going to work in the space of the output files, ie: write out pngs in sRGB, then use those pngs in some other software that is compositing natively in sRGB.  This is the general assumption for how pngs are used, but it's also true that maybe we don't care about anyone working in any space other than linear.

The focus on roundtripping feels reasonable, and the big win in functionality is the alpha case being obviously improved.  Disabling the unpremult for images without alpha does seem reasonable.  To implement this, I would have to add a new context variable, similar to `__imageWriter:colorSpace`, but something like `__imageWriter:hasAlpha`, which could be set based on both the channel mask and the file format?

johnhaddon commented:
> To implement this, I would have to add a new context variable, similar to __imageWriter:colorSpace, but something like __imageWriter:hasAlpha, which could be set based on both the channel mask and the file format?

That would be one way, but I think it'd be fiddly. The `__imageWriter:colorSpace` thing works nicely because we can use string substitutions on the colorspace plug - the same doesn't work for a BoolPlug, so we'd need a little BoolContextVariable utility node. Would an internal DeleteChannels node, in Keep mode, with the channels being driven by `ImageWriter.channels` work? Or in Delete mode, with an `__imageWriter:channelsToDelete` context variable?

danieldresser-ie commented:
> Would an internal DeleteChannels node, in Keep mode, with the channels being driven by ImageWriter.channels

That would be a much simpler and nicer solution, but it doesn't really allow us to drive it based on the file format, does it?

> Or in Delete mode, with an __imageWriter:channelsToDelete context variable?

Bit messier , but  I guess that does allow us to drive it based on the file format?

johnhaddon commented:
>> Would an internal DeleteChannels node, in Keep mode, with the channels being driven by ImageWriter.channels

> That would be a much simpler and nicer solution, but it doesn't really allow us to drive it based on the file format, does it?

Yeah, I just realised that. It works for my "have RGBA, but writing RGB" case, but not for the "writing RGBA but file format discards alpha" case.

>> Or in Delete mode, with an __imageWriter:channelsToDelete context variable?

> Bit messier , but I guess that does allow us to drive it based on the file format?

Yep, that should work. Couple of other alternatives :

- Two internal ColorSpace nodes, one with and one without `processUnpremultiplied` turned on.
- Use the fact that you can make a StringPlug -> BoolPlug connection, with empty strings being converted to false. You could then feed a context variable into the string plug, and have it convert to bool automatically.

Just so we're clear, you are now convinced we should treat "have RGBA, but writing RGB" and "have RGBA, but A not supported by file format" the same way?

danieldresser-ie commented:
> Two internal ColorSpace nodes, one with and one without processUnpremultiplied turned on.

Oh that, seems simplest.  Lets go with that.

> Just so we're clear, you are now convinced we should treat "have RGBA, but writing RGB" and "have RGBA, but A not supported by file format" the same way?

Yeah.  Roundtripping in Gaffer is persuasive.  The case that fails then is exporting channels in separate files to software that does blending in non-linear colorspace, and that's not really something we care about.

danieldresser-ie commented:
OK, I've added the separate path for alpha channels which are not being written due to channel masks or file format restrictions.  I think this is looking pretty good now.

I added a quick test for roundtripping files with the alpha removed - the required error threshold of 0.011 is a little high, but I think it makes sense that writing back and forth to an 8 bit format introduces some noticeable error.  The test is adequate to show that it's basically working ( on master it fails with an error of 0.28 due to the clamping issue I found - without the alpha it works properly, but the expected images on master are bad due to clamping )

Oh, and I recreated your commit - it was quicker to download the text through github than getting the PR pulled back through.

johnhaddon commented:
Great - thanks Daniel!
